torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -449,6 +474,90 @@ void ProcessGroupAgent::listenLoop() {
   }
 }
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  while (!shutdown_.load()) {
+    std::chrono::milliseconds sleepTime;
+    {
+      std::unique_lock<std::mutex> lock{futureMutex_};
+      // Estimate amount of time the first future will time out in, and sleep
+      // for that long.
+      // if there are no futures or the RPC timeout is set to 0 (meaning no
+      // timeout), then sleep for a set "infinity" time.
+      if (futureTimeouts_.empty() || rpcTimeout_.count() == 0) {
+        sleepTime = INFINITY_SLEEP_INTERVAL;
+      } else {
+        auto runningTime =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::steady_clock::now().time_since_epoch()) -
+            futureTimeouts_.begin()->first;
+        sleepTime =
+            std::max(rpcTimeout_ - runningTime, std::chrono::milliseconds(0));
+      }
+      futureTimeoutCV_.wait_for(lock, sleepTime);
+    }
+
+    if (shutdown_.load()) {
+      return;
+    }
+
+    const auto timedOutFutures = processTimedOutFutures();
+
+    // Do not hold the lock while marking futures completed, as markCompleted()
+    // could invoke callbacks.
+    if (!timedOutFutures.empty()) {
+      std::stringstream ss;
+      ss << "Future ran for more than " << rpcTimeout_.count()
+         << " milliseconds and timed out.";
+      const auto exceptionMsg = createExceptionResponse(
+          Message({}, {}, MessageType::EXCEPTION), ss.str());
+      for (const auto& timedOutFuture : timedOutFutures) {
+        timedOutFuture.future_->markCompleted(exceptionMsg);
+
+        const int dst = timedOutFuture.dstRank_;
+        recvCounts_.increment(dst);
+        futureCV_.notify_all();
+      }
+    }
+  }
+}
+
+const std::vector<ProcessGroupAgent::FutureInfo> ProcessGroupAgent::
+    processTimedOutFutures() {
+  std::vector<FutureInfo> timedOutFutures;
+  {
+    // Check for timed out futures while holding the lock.
+    std::lock_guard<std::mutex> lock{futureMutex_};
+    for (auto it = futureTimeouts_.begin(); it != futureTimeouts_.end();
+         /* intentional no increment */) {
+      const std::chrono::milliseconds startTime = it->first;
+      const std::vector<int64_t> futureIDs = it->second;
+      const auto diffTime =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              std::chrono::steady_clock::now().time_since_epoch()) -
+          startTime;
+      if (rpcTimeout_.count() == 0 || diffTime <= rpcTimeout_) {
+        // Since the futureTimeouts_ map is ordered by timeout, we don't need
+        // to check the remaining futures.
+        break;
+      } else {
+        for (const auto& futureID : futureIDs) {
+          auto it = futures_.find(futureID);
+          TORCH_CHECK(
```

Comment:
nit: TORCH_INTERNAL_ASSERT, TORCH_CHECK is mostly to indicate that the user made some error.